### ISSUE TITLE
Improving efficiency of Stan code

### ISSUE THREAD

**Issue by ScottClaessens** [ScottClaessens/coevolve#27]
I posted the model code over on the Stan forums (see [here](https://discourse.mc-stan.org/t/improving-efficiency-for-dynamic-coevolutionary-stan-model/36103)). Bob Carpenter and Aki Vehtari had some good advice for improving the efficiency of the model. I'm opening this issue so we can work on implementing some of their suggestions.

**Comment by ScottClaessens** [comment]
I ran some basic profiling on the Stan code for two Bernoulli variables, as recommended by Bob Carpenter. The profiling suggests that the `cov_drift()` function is taking the most time to run. This is where we should focus our optimisation efforts.

```
          name thread_id total_time forward_time reverse_time chain_stack no_chain_stack autodiff_calls no_autodiff_calls
1    ancestral         1  0.0055716    0.0039069    0.0016647           0         123020          30755              1001
2 calculateEta         1  1.1834200    0.8167090    0.3667080     1568505        4982310         246040              8008
3     cholesky         1  0.3611990    0.2162700    0.1449290      492080        1476240         246040              8008
4    cov_drift         1  5.0586500    3.4009100    1.6577400   288322508       43691860         246040              8008
5   likelihood         1  0.0868738    0.0817592    0.0051146      615100         307550          30755                 1
6       priors         1  0.0216827    0.0176221    0.0040606      184530         184530          30755                 1
```

Here's the Stan code with profile statements:

```stan
// Generated with coevolve 0.0.0.9010
functions {
  // returns the Kronecker Product
  matrix kronecker_prod(matrix A, matrix B) {
    matrix[rows(A) * rows(B), cols(A) * cols(B)] C;
    int m;
    int n;
    int p;
    int q;
    m = rows(A);
    n = cols(A);
    p = rows(B);
    q = cols(B);
    for (i in 1:m)
      for (j in 1:n)
        for (k in 1:p)
          for (l in 1:q)
            C[p*(i-1)+k,q*(j-1)+l] = A[i,j]*B[k,l];
    return C;
  }

  // expected auto and cross effects over a discrete time t
  matrix A_dt(matrix A, real t) {
    return( matrix_exp(A * t) );
  }

  // calculate A sharp matrix
  matrix A_sharp(matrix A) {
    matrix[rows(A) * rows(A), cols(A) * cols(A)] A_temp;
    matrix[rows(A),cols(A)] I; // identity matrix
    I = diag_matrix(rep_vector(1,rows(A)));
    A_temp = kronecker_prod(A,I) + kronecker_prod(I,A);
    return(A_temp);
  }

  // solve SDE
  matrix cov_drift(matrix A, matrix Q, real ts) {
    matrix[rows(A) * rows(A), cols(A) * cols(A)] A_sharp_temp;
    matrix[rows(A) * rows(A), cols(A) * cols(A)] I; // identity matrix
    vector[rows(Q)*cols(Q)] row_Q;
    vector[rows(A)*cols(A)] irow_vec;
    matrix[rows(A),cols(A)] irow_mat;
    I = diag_matrix(rep_vector(1,rows(A_sharp_temp)));
    A_sharp_temp = A_sharp(A);
    // row operation takes elements of a matrix rowwise and puts them into a column vector
    for (i in 1:rows(Q))
      for (j in 1:cols(Q)) {
        row_Q[i + (j-1)*rows(Q)] = Q[j,i];
      }
    irow_vec = inverse(A_sharp_temp) * (matrix_exp(A_sharp_temp * ts) - I) * row_Q;
    // irow takes elements of a column vector and puts them in a matrix rowwise
    {
      int row_size = rows(A);
      int row_ticker = 1;
      int col_ticker = 0;
      for (i in 1:num_elements(irow_vec)) {
        col_ticker += 1;
        if (col_ticker > row_size) {
          row_ticker += 1;
          col_ticker = 1;
        }
        irow_mat[row_ticker,col_ticker] = irow_vec[i];
      }
    }
    return(irow_mat);
  }

  // return number of matches of y in vector x
  int num_matches(vector x, real y) {
    int n = 0;
    for (i in 1:rows(x))
      if (x[i] == y)
        n += 1;
    return n;
  }

  // return indices in vector x where x == y
  array[] int which_equal(vector x, real y) {
    array [num_matches(x, y)] int match_positions;
    int pos = 1;
    for (i in 1:rows(x)) {
      if (x[i] == y) {
        match_positions[pos] = i;
        pos += 1;
      }
    }
    return match_positions;
  }
}
data{
  int N_tips; // number of tips
  int N_obs; // number of observations
  int J; // number of response traits
  int N_seg; // total number of segments in the tree
  array[N_seg] int node_seq; // index of tree nodes
  array[N_seg] int parent; // index of the parent node of each descendent
  array[N_seg] real ts; // time since parent
  array[N_seg] int tip; // indicator of whether a given segment ends in a tip
  array[J,J] int effects_mat; // which effects should be estimated?
  int num_effects; // number of effects being estimated
  matrix[N_obs,J] y; // observed data
  matrix[N_obs,J] miss; // are data points missing?
  array[N_obs] int tip_id; // index between 1 and N_tips that gives the group id
  int prior_only; // should the likelihood be ignored?
}
transformed data{
  vector[to_int(N_obs - sum(col(miss, 1)))] obs1; // observed data for variable 1
  vector[to_int(N_obs - sum(col(miss, 2)))] obs2; // observed data for variable 2
  obs1 = col(y, 1)[which_equal(col(miss, 1), 0)];
  obs2 = col(y, 2)[which_equal(col(miss, 2), 0)];
}
parameters{
  vector<upper=0>[J] A_diag; // autoregressive terms of A
  vector[num_effects - J] A_offdiag; // cross-lagged terms of A
  vector<lower=0>[J] Q_diag; // self-drift terms
  vector[J] b; // SDE intercepts
  vector[J] eta_anc; // ancestral states
  matrix[N_seg - 1,J] z_drift; // stochastic drift, unscaled and uncorrelated
}
transformed parameters{
  matrix[N_seg,J] eta;
  matrix[J,J] Q; // drift matrix
  matrix[J,J] I; // identity matrix
  matrix[J,J] A; // selection matrix
  vector[J*J - J] Q_offdiag = rep_vector(0.0, J*J - J);
  matrix[N_seg,J] drift_tips; // terminal drift parameters
  matrix[N_seg,J] sigma_tips; // terminal drift parameters
  // fill A matrix //////////
  {
    int ticker = 1;
    // fill upper triangle of matrix
    for (i in 1:(J-1)) {
      for (j in (i+1):J) {
        if (effects_mat[i,j] == 1) {
          A[i,j] = A_offdiag[ticker];
          ticker += 1;
        } else if (effects_mat[i,j] == 0) {
          A[i,j] = 0;
        }
      }
    }
    // fill lower triangle of matrix
    for (i in 1:(J-1)) {
      for (j in (i+1):J) {
        if (effects_mat[j,i] == 1) {
          A[j,i] = A_offdiag[ticker];
          ticker += 1;
        } else if (effects_mat[j,i] == 0) {
          A[j,i] = 0;
        }
      }
    }
    // fill diag of matrix
    for (j in 1:J) A[j,j] = A_diag[j];
  }
  // fill Q matrix //////////
  {
    int ticker = 1;
    for (i in 1:(J-1))
      for (j in (i+1):J) {
        Q[i,j] = Q_offdiag[ticker];
        Q[j,i] = Q[i,j]; // symmetry of covariance
        ticker += 1;
      }
    for (j in 1:J) Q[j,j] = Q_diag[j];
  }
  // identity matrix
  I = diag_matrix(rep_vector(1,J));
  // setting ancestral states and placeholders
  profile("ancestral") {
    for (j in 1:J) {
      eta[node_seq[1],j] = eta_anc[j];
      drift_tips[node_seq[1],j] = -99;
      sigma_tips[node_seq[1],j] = -99;
    }
  }
  for (i in 2:N_seg) {
    matrix[J,J] A_delta; // amount of deterministic change (selection)
    matrix[J,J] VCV; // variance-covariance matrix of stochastic change (drift)
    vector[J] drift_seg; // accumulated drift over the segment
    A_delta = A_dt(A, ts[i]);
    profile("cov_drift") {VCV = cov_drift(A, Q, ts[i]);}
    profile("cholesky") {drift_seg = cholesky_decompose(VCV) * to_vector( z_drift[i-1,] );}
    // if not a tip, add the drift parameter
    profile("calculateEta") {
      if (tip[i] == 0) {
        eta[node_seq[i],] = to_row_vector(
          A_delta * to_vector(eta[parent[i],]) + (inverse(A) * (A_delta - I) * b) + drift_seg
        );
        drift_tips[node_seq[i],] = to_row_vector(rep_vector(-99, J));
        sigma_tips[node_seq[i],] = to_row_vector(rep_vector(-99, J));
      }
      // if is a tip, omit, we'll deal with it in the model block;
      else {
        eta[node_seq[i],] = to_row_vector(
          A_delta * to_vector(eta[parent[i],]) + (inverse(A) * (A_delta - I) * b)
        );
        drift_tips[node_seq[i],] = to_row_vector(drift_seg);
        sigma_tips[node_seq[i],] = to_row_vector(diagonal(Q));
      }
    }
  }
}
model{
  profile("priors") {
    b ~ std_normal();
    eta_anc ~ std_normal();
    to_vector(z_drift) ~ std_normal();
    A_offdiag ~ std_normal();
    A_diag ~ std_normal();
    Q_diag ~ std_normal();
  }
  profile("likelihood") {
    if (!prior_only) {
    for (i in 1:N_obs) {
        if (miss[i,1] == 0) to_int(y[i,1]) ~ bernoulli_logit(eta[tip_id[i],1] + drift_tips[tip_id[i],1]);
        if (miss[i,2] == 0) to_int(y[i,2]) ~ bernoulli_logit(eta[tip_id[i],2] + drift_tips[tip_id[i],2]);
    }
  }
  }
}
```

And here's the R code for running this:

```r
library(coevolve)
library(cmdstanr)
set.seed(1)
# simulate data
n <- 5
tree <- ape::rcoal(n)
d <- data.frame(
  id = tree$tip.label,
  x = rbinom(n, 1, 0.5),
  y = rbinom(n, 1, 0.5)
)
# get data for stan model
sd <-
  coev_make_standata(
    data = d,
    variables = list(
      x = "bernoulli_logit",
      y = "bernoulli_logit"
    ),
    id = "id",
    tree = tree
  )
# fit stan model with profiling
model <- cmdstan_model(stan_file = "profiling.stan")
fit <- model$sample(data = sd, chains = 1)
# get profiling info
fit$profiles()
```

**Comment by ScottClaessens** [comment]
I've replaced the Kronecker sum function with Erik's `kronecker_sum_self()` function in b1c6714. Erik found that this was faster when the number of traits increases beyond two.

**Comment by ScottClaessens** [comment]
Unfortunately, the model fails with the `kronecker_sum_self()` function. The model compiles fine and is syntactically correct, but crashes with the following error upon fitting:

```
Chain 1 Rejecting initial value:
Chain 1   Error evaluating the log probability at the initial value.
Chain 1 Exception: cholesky_decompose: m is not symmetric. m[1,2] = 9.30691e+12, but m[2,1] = 5.68581e+12 (in 'C:/Users/scla896/AppData/Local/Temp/Rtmpea4jYE/model-4fbc6c287040.stan', line 176, column 4 to column 69)
```

I have reverted to the previous function in 18cf1e6.

**Comment by ScottClaessens** [comment]
I've done some more line-by-line profiling of the `cov_drift()` function in the Stan code. The profiling suggests that our biggest bottleneck in the function is the `irow_vec` line:

```stan
// solve SDE
matrix cov_drift(matrix A, matrix Q, real ts) {
  matrix[rows(A) * rows(A), cols(A) * cols(A)] A_sharp_temp;
  vector[rows(Q)*cols(Q)] row_Q;
  vector[rows(A)*cols(A)] irow_vec;
  matrix[rows(A),cols(A)] irow_mat;
  profile("kronecker_sum") {A_sharp_temp = kronecker_sum(A, A);}
  // row operation takes elements of a matrix rowwise and puts them into a column vector
  profile("row_Q") {
    for (i in 1:rows(Q))
      for (j in 1:cols(Q)) {
        row_Q[i + (j-1)*rows(Q)] = Q[j,i];
      }
  }
  profile("irow_vec") {irow_vec = (A_sharp_temp \ add_diag(matrix_exp(A_sharp_temp * ts), -1)) * row_Q;}
  // irow takes elements of a column vector and puts them in a matrix rowwise
  profile("irow_mat") {
    {
      int row_size = rows(A);
      int row_ticker = 1;
      int col_ticker = 0;
      for (i in 1:num_elements(irow_vec)) {
        col_ticker += 1;
        if (col_ticker > row_size) {
          row_ticker += 1;
          col_ticker = 1;
        }
        irow_mat[row_ticker,col_ticker] = irow_vec[i];
      }
    }
  }
  return irow_mat;
}
```

```
           name thread_id total_time forward_time reverse_time chain_stack no_chain_stack autodiff_calls no_autodiff_calls
1      irow_mat         1  0.0639180    0.0359620    0.0279560           0              0         589068             18018
2      irow_vec         1  9.4457300    5.9759000    3.4698300   534138035       59645256         589068             18018
3 kronecker_sum         1  0.7651890    0.5353950    0.2297940    37700416         589069         589068             18018
4         row_Q         1  0.0646981    0.0367919    0.0279062           0              0         589068             18018
```

@erik-ringen is there anything we can do to simplify this line?

**Comment by ScottClaessens** [comment]
See Charles Driver's response here: https://discourse.mc-stan.org/t/improving-efficiency-for-dynamic-coevolutionary-stan-model/36103/8?u=scott.claessens

**Comment by ScottClaessens** [comment]
And his follow-up here: https://discourse.mc-stan.org/t/improving-efficiency-for-dynamic-coevolutionary-stan-model/36103/9

This is super useful! @erik-ringen do you think we could implement something like this? If I'm understanding right, Charles is suggesting that we do the covariance computation just once, rather than repeat it for every segment of the tree.

**Comment by erik-ringen** [comment]
Yes, it helped me to re-read his 2018 handbook article to understand the logic of the asymptotic covariance calculation. Below are the updates to the Stan code:

For the `functions` block:

`
// Charles Driver's optimized way of solving for the asymptotic Q matrix
  matrix ksolve(matrix A, matrix Q){
    int d= rows(A);
    int d2= (d*d-d)%/%2;
    matrix[d+d2,d+d2] O;
    vector[d+d2] triQ;
    matrix[d,d] AQ;
    int z=0; //z is row of output
    for(j in 1:d){//for column reference of solution vector
      for(i in 1:j){ //and row reference...
        if(j >= i){ //if i and j denote a covariance parameter (from upper tri)
          int y=0; //start new output row
          z+=1; //shift current output row down
          
          for(ci in 1:d){//for columns and
            for(ri in 1:d){ //rows of solution
              if(ci >= ri){ //when in upper tri (inc diag)
                y+=1; //move to next column of output
                
                if(i==j){ //if output row is for a diagonal element
                  if(ri==i) O[z,y] = 2*A[ri,ci];
                  if(ci==i) O[z,y] = 2*A[ci,ri];
                }
                
                if(i!=j){ //if output row is not for a diagonal element
                  if(y==z) O[z,y] = A[ri,ri] + A[ci,ci]; //if column of output matches row of output, sum both A diags
                  if(y!=z){ //otherwise...
                    // if solution element we refer to is related to output row...
                    if(ci==ri){ //if solution element is a variance
                      if(ci==i) O[z,y] = A[j,ci]; //if variance of solution corresponds to row of our output
                      if(ci==j) O[z,y] = A[i,ci]; //if variance of solution corresponds to col of our output
                    }
                    if(ci!=ri && (ri==i||ri==j||ci==i||ci==j)){//if solution element is a related covariance
                      //for row 1,2 / 2,1 of output, if solution row ri 1 (match) and column ci 3, we need A[2,3]
                      if(ri==i) O[z,y] = A[j,ci];
                      if(ri==j) O[z,y] = A[i,ci];
                      if(ci==i) O[z,y] = A[j,ri];
                      if(ci==j) O[z,y] = A[i,ri];
                    }
                  }
                }
                if(is_nan(O[z,y])) O[z,y]=0;
              }
            }
          }
        }
      }
    }
    
    z=0; //get upper tri of Q
    for(j in 1:d){
      for(i in 1:j){
        z+=1;
        triQ[z] = Q[i,j];
      }
    }
    triQ=-O \ triQ; //get upper tri of asymQ
    
    z=0; // put upper tri of asymQ into matrix
    for(j in 1:d){
      for(i in 1:j){
        z+=1;
        AQ[i,j] = triQ[z];
        if(i!=j) AQ[j,i] = triQ[z];
      }
    }
    return AQ;
  }
`

For the `transformed parameters` block:

`
matrix[J, J] Q_inf; // asymptotic covariance matrix
...
Q_inf = ksolve(A, Q);
...
VCV = Q_inf - quad_form_sym(Q_inf, A_delta');

`

The explicit kronecker product functions can be removed now, I think. In very limited testing I have seen a speed-up of 2.5x!

**Comment by ScottClaessens** [comment]
Great!  I'll look at adding this to the Stan code soon.

**Comment by ScottClaessens** [comment]
@erik-ringen could you email me a copy of the full Stan code with these additions?

**Comment by ScottClaessens** [comment]
I implemented the new function, along with some other optimisations, in af0b237. I have implemented most of what Bob Carpenter suggested, though some I tried to implement but my changes kept breaking the model so I left them in the end.

I think the most important thing is the huge speed up we are seeing with the new version of the model. The new model in the README, for example, took 28% of the wall time compared to the old model!

@erik-ringen it would be great if you could skim your eyes over the Stan code that is produced by `coev_make_stancode()` on the optimisation branch. If you can't see any glaring errors, then we can merge this into the main branch and close this issue.

**Comment by ScottClaessens** [comment]
I merged the optimisation branch in 72db79d.
